contracts/d3vent.sol

```diff
@@ -0,0 +1,165 @@
+    //     address userAddress;
+    //     event_[] eventIds;  // do we need quick lookups and/or able to iterate eventIds?
+    // }
+
+    struct event_ {
+        address organiser;
+        uint id;
+        string name;
+        string uri;
+        uint dateTime;
+        uint price;
+        uint128 capacity;
+        uint128 numJoined;      
+        bool isJoinable;
+        uint withdrawalDate;
+    }
+
+    event_[] events;
+
+
+    mapping(uint => mapping(address => bool)) isJoined;    // eventId => userAddr => isJoined
+    mapping(uint => mapping(address => uint)) joinerBalances;   // eventId => userAddr => balance
+    mapping(uint => uint) eventBalances;    // eventId => event balance
+    mapping(address => bool) isAdmin;   // admin address to bool
+    
+
+    constructor (uint _withdrawalBuffer) {
+        withdrawalBuffer = _withdrawalBuffer;
+    }
+    
+    modifier isAuthorised (uint _eventId) {
+        require(
+            msg.sender == events[_eventId].organiser || 
+            isAdmin[msg.sender], 
+            "not authroised"
+        );
+        _;
+    }
+
+    // test parameters: "test event name","https://livepeer.org/123",1662994265,1000,1000000,false
+    function createEvent(
+        //string calldata _eventName,
+        string calldata _name,
+        string calldata _uri,
+        uint _dateTime,
+        uint _price,
+        uint128 _capacity,
+        bool _isJoinable
+        ) external {              // public to allow internal calls for testing
+
+        // sanity checks
+        require(_dateTime > block.timestamp, "date/time in past");
+
+        event_ memory newEvent;
+        
+        newEvent.organiser = msg.sender;
+        newEvent.id = eventIds++;
+        newEvent.name = _name;
+        newEvent.uri = _uri;
+        newEvent.dateTime = _dateTime;
+        newEvent.price = _price;
+        newEvent.capacity = _capacity;
+        newEvent.isJoinable = _isJoinable;
+    
+        events.push(newEvent);
+        emit CreatedEvent(newEvent.id, newEvent.dateTime, newEvent.name);
+    }
+
+
+    function setOrganiser(uint _eventId, address _newOrganiser) external {
+        // sanity checks
+        require(_newOrganiser != address(0), "invalid: zero address");
+        require(_eventId <= eventIds, "invalid event id");
+        events[_eventId].organiser = _newOrganiser;
+        // emit event
+    }
+
+
+    function setEventIsJoinable(uint _id, bool _isJoinable) external {
+        require(! events[_id].isJoinable, "already joinable");
+        events[_id].isJoinable = _isJoinable;
+        emit JoinableSet(_id, _isJoinable);
+    }
+
+    // @dev needs proof of personhood
+    // support superfluid payments? how would that affect logic
+    // 
+    function joinEvent(uint _id) external payable {
+        require(_id <= eventIds, "invalid event id");
+        require(events[_id].isJoinable, "cant join at this time");
+        require(msg.value == events[_id].price, "send event price");
```

Comment:
msg.value >= events[_id].price. We can't expect user to always send the exact amount. Amount greater than the price is just fine I think.

Reply:
we can accept overpayments, but why couldn't we expect the user to send the exact price? The front end should set the price correctly in the users wallet. How would they end up paying too much?

Reply:
User will always send the exact amount but accepting overpayments won't harm anything. I think it's just a good practice!